<a href="https://colab.research.google.com/github/lblogan14/reinforcement_learning_with_tensorflow/blob/master/ch3_MDP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
%cd /content/drive/My' 'Drive/Colab' 'Notebooks/Reinforcement_Learning_with_TensorFlow/

/content/drive/My Drive/Colab Notebooks/Reinforcement_Learning_with_TensorFlow


#Markov Decision Processes
The Markov decision process (MDP) is a reinforcement learning approach in a gridworld environment containing sets of states, actions, and rewards, following the Markov property to obtain an optimal policy.

MDP is defined as the collection of:
* **States**: $S$
* **Actions**: $A(s), A$
* **Transition model**: $T(s,a,s')\sim P(s'|s,a)$
* **Rewards**: $R(s), R(s,a), R(s,a,s')$
* **Policy**: $\pi(s)\rightarrow a_{\pi^*}$ is the optimal policy

The MDP environment is fully observable, which means whatever observation the agent makes at any point in time is enough to make an optimal decision.

For partially observable environment, the agent needs a memory to store the past observations to make the best possible decisions.

##Markov property
The present information at time $t$ matters in order to know the information of near future at time $t+1$. 

*First Order* of Markov: $x_t$ depends only on $x_{t-1}$.
$$P(x_t|x_{t-1},x_{t-2},...,x_t)=P(x_t|x_{t-1})$$

*Second Order* of Markov: $x_t$ depends only on $x_{t-1}$ and $x_{t-2}$.
$$P(x_t|x_{t-1},x_{t-2},...,x_t)=P(x_t|x_{t-1}, x_{t-2})$$.

Most of the RL algorithms are based on the first order of Markov property.

##$S$ State Set
is a set of different states, represented as $s$ constituting the envrionment. States are the feature representation of the data obtained from the environment. State spaces can be discrete or continuous.

Consider a gridworld having 12 discrete states:
![alt text](https://github.com/lblogan14/reinforcement_learning_with_tensorflow/blob/master/note_images/ch3/gridworld.PNG?raw=true)

The states can be represented as 1, 2,..., 12 or by coordinates, (1, 1), (1, 2), ...., (3, 4)


##Actions
are the choices an agent can perform or execute in a particular state. actions can also be discrete or continuous. For the gridworld shown above, it has 12 discrete states and 4 discrete actions which are *UP*, *DOWN*, *RIGHT*, and *LEFT*.

The action space can be represented as $a\in A$, where $A=\{UP, DOWN, RIGHT, LEFT\}$. It can also be represented as a function of state, $a=A(s)$, where depending on the state function, it decides which action is possible.

##Transition model
is defined as a function of three variables $T(s,a,s')$, which consists of the *current state* $s$, *action* $a$, and the *new state* $s'$. The transition model also defines the rules to play game in the envirnoment and gives probability $P(s'|s,a)$, the probability of landing up in the new $s'$ state given that the agent takes an action, $a$, in given state, $s$.

Still consider the gridworld example,
![alt text](https://github.com/lblogan14/reinforcement_learning_with_tensorflow/blob/master/note_images/ch3/gridworld.PNG?raw=true)

This environment can be considered as*determined* or *stochastic*:
* **Determined environment**: The certain action will be certainly performed with probability 1.
* **Stochastic environment**: The same action will ben taken with a probability less than 1 and there are other probable actions to select.

##Rewards
quantifies the usefulness of entering into a state. The forms to represent the reward are: $R(s), R(s,a), R(s,a,s')$, equivalently.

Two approaches to reward the agent for when taking a certain action:
* **Credit assignment problem**: look at the past and check which actions led to the present reward. Find which action gets the credit
* **Delayed rewards**: check which action to take in the present state will lead to potential rewards.

##Policy
is the solution to the MDP problem.
$$\mathrm{Policy}:\pi(s)\rightarrow a$$
The policy is a function taking the state as an input and outputs the action to be taken, so the policy is a command that the agent has to obey.

$\pi^*$ is the optimal policy which maximizes the expect reward.

##Sequence of rewards
is used to find the optimal policy for MDP problem, but there are assumptions for a sequence of rewards applied on the concept of delayed rewards.

###Assumption 1: The inifinite horizons
The infinite amount of time steps to reach goal state from start state. Thus, $\pi(s)\rightarrow a$ so the policy function does not take the remaining time steps. If it had been a finite horizon, $\pi(s,t)\rightarrow a$, where $t$ is the time steps left to get the task done.

###Assumption 2: Utility of sequences
The utility of sequences refers to the overall reward received when the agent goes through the sequences of states, defined as $U(s_0, s_1, s_2,...)$, where $s_0, s_1, s_2,...$ represents the sequence of states.

This assumption is that if there are two utilities, $U(s_0, s_1, s_2,....)$ and $U(s_0, s'_1, s'_2,....)$ with the same start state, and 
$$U(s_0, s_1, s_2,....)>U(s_0, s'_1, s'_2,....)$$
then,
$$U(s_1, s_2,....)>U(s'_1, s'_2,....)$$
This called the stationary of preferences.

The following equation implements the concept of delayed rewards and satisfies the stationary of preferences:
$$U(s_0,s_1,s_2,...)=\sum\limits_{t=0}^\infty \gamma^t R(s_t)$$
where $0\leq\gamma\leq 1$

This summation of all the rewards $R(s_t)$ has a upper bound,
$$\sum\limits_{t=0}^\infty \gamma^t R(s_t) \leq \frac{R_\max}{1-\gamma}$$
where $R_\max$ is the maximum reward value

##Bellman equations
$$\pi^*=\arg\max_\pi E[\sum\limits_{t=0}^\infty \gamma^t R(s_t)|\pi]$$
where $E[\sum\limits_{t=0}^\infty \gamma^t R(s_t)|\pi]$ means the expected value of the rewards obtained from the sequence of states agent observes if it follows the $\pi$ policy. In addition, $\arg\max_\pi$ outputs the $\pi$ policy that has the highest expected reward.

**Utility of the policy of a state**, $U^\pi(s)$ at the $s$ state, given a $\pi$ policy, is the expected rewards from that state onward:
$$U^\pi(s)=E[\sum\limits_{t=0}^\infty \gamma^t R(s_t)|\pi, s_0=s]$$

Immediate reward of the state: $R(s)$

Optimal policy: policy that maximizes the expected utility
$$\pi^*=\arg\max_a\sum\limits_{s'}T(s,a,s')U(s')$$
where $T(s,a,s')$ is the transition probability, $P(s'|s,a)$ and $U(s')$ is the utility of the new landing state after the $a$ action is taken on the $s$ state.

The utility of the $s$ state is given by the Bellman equation:
$$U(s)=R(s)+\gamma\max_a\sum\limits_{s'}T(s,a,s')U(s')$$
where
* $R(s)$ is the immediate reward
* $\max_a\sum\limits_{s'}T(s,a,s')U(s')$ is the reward from future, discounted utilities of the $s$ state where the agent can reach from the given $s$ state if the action $a$ is taken.

###Solve the Bellman equation to find policies
Suppose $n$ states are given in the environment and the Bellman equation, thus, there are $n$ equations and $n$ unknowns.
To solve this system of equations,
1. Start with an arbitrary utility
2. Update the utilities based on the neighborhood until convergence, that is, update
the utility of the state using the Bellman equation based on the utilities of the
landing states from the given state

####Value iteration
iterates multiple times to convergence towards the true value of the states.
####Policy iteration
iterates over the policy and updates the policy itself instead of value until the policy converges to the optimum.

Process of policy iteration:
* Start with a random policy, $\pi_0$
* For the given $\pi_t$ policy at iteration step $t$, calculate $U_t=U_t^\pi$ using:
$$U_t(s)=R(s)+\gamma\max_a\sum\limits_{s'}T(s,a,s')U_{t-1}(s')$$
* Improve the $\pi_{t+1}$ policy by
$$\pi_{t+1}=\arg\max_a\sum\limits_{s'}T(s,a,s')U_t(s')$$

#Partially Observable Markov Decision Processes
In MDP, the observable quantities are action, set $A$, the state, set $S$, transition model, $T$,
and rewards, set $R$.

In **partially observable MDP**, a.k.a, **POMDP**, some of the above quantities are not directly observable to the agent.

In POMDP, there is an observation set, $Z$, containing different observable states and a observation function, $O$, which takes the $s$ state and the $z$ observation as inputs and outputs the probability of seeing that $z$ observation in the $s$ state.

In summary,
* **MDP**: $\{S,A,T,R\}$
* **POMDP**: $\{S,A,Z,T,R,O\}$

where $S,A,T,R$ are the same.

For a POMDP to be a true MDP, the following conditions must be satisfied:

* $Z=S$, that is, fully observe all states \\
* $
O(s,z) = \left\{
        \begin{array}{ll}
            1 & \quad s=z \\
            0 & \quad \mathrm{otherwise}
        \end{array}
    \right.
$

POMDP are hugely intractable to solve optimally.

##State estimation
Expansion of the stat spaces helps convert the POMDP into an MDP where $Z$ contains fully observable state space.

The concept of the **belief state**, $b(s)$, is introduced here, which is the state that the decision maker uses in the context of a POMDP. This belief state, $b(s)$, gives the probability of the agent being in the $s$ state. Thus, the belief state, $b$, is a vector representing the probability distribution over all states and it get updated once an action is taken.

For example, \\
there is a belief state, $b$, the agent takes an action, $a$, and some observations, $z$, received, then this will form a new belief state. Therefore, the new belief state, $b'(s')$,
$$b'(s')=\mathrm{probability\,of\, being\,in}\,s\,\mathrm{state\,given\,after}\,b,a,z,\,\mathrm{that\,is,}\,p(s'|b,a,z)$$
 \\
$$p(s'|s,a,z)=\frac{p(z|b,a,s')}{p(z|b,a)}\sum\limits_{s}b(s)\cdot p(s'|s,b,a)$$
where, $p(z|b,a,s')=O(s',z)$ and $p(s'|s,b,a)=T(s,a,s')$. Therefore,
$$p(s'|s,a,z)=\frac{O(s',z)}{p(z|b,a)}\sum\limits_{s}b(s)\cdot T(s,a,s')$$

##Value iteration in POMDPs
Value iteration on an infinite state space obtained from a belief MDP:
* $V_0(b)=0$ at $t=0$
* $V_t(b)=\max_{a}[R(b,a)+\gamma\sum\limits_{z}P(z|b,a)\cdot V_{t-1}(b')]$ at $t>0$, where $b'$ is $b'(s')=p(s'|b,a,z)$, which is the state estimation for $(b,a,z)$ and $R(b,a)$ is the expected reward over a belief state:
$$R(b,a)=\sum\limits_{s}p(s)\cdot R(s,a)=\sum\limits_{s}b(s)\cdot R(s,a)$$
where, \\
$p(s)=$ probability of the $s$ state \\
$R(s,a)=$ reward in that state \\
$\sum\limits_{s}b(s)\cdot R(s,a)=$ expected reward over a belief state

#Train `FrozenLake-v0` using MDP

In [0]:
# dependent libraries
from __future__ import print_function
import gym
import numpy as np
import time

In [2]:
# load environment
env = gym.make('FrozenLake-v0')

/usr/local/lib/python3.6/dist-packages/gym/envs/registration.py:14: PkgResourcesDeprecationWarning: Parameters to load are deprecated.  Call .resolve and .require separately.
  result = entry_point.load(False)


In [3]:
s = env.reset()
print(s)

0


In [4]:
env.render()


SFFF
FHFH
FFFH
HFFG


In [5]:
print(env.action_space) # number of actions
print(env.observation_space) # number of states

Discrete(4)
Discrete(16)


In [6]:
print("Number of actions : ",env.action_space.n)
print("Number of states : ",env.observation_space.n)

Number of actions :  4
Number of states :  16


In [7]:
# ## Value Iteration Implementation

# initialize utilities of all states with zeros
U = np.zeros([env.observation_space.n])

# terminal states have utility values equal to their reward
U[15] = 1 # goal states
U[[5,7,11,12]] = -1 # hole states
termS = [5,7,11,12,15] # terminal states

# set hyperparameters
y = 0.8 # discounted factor
eps = 1e-3 # threshold if the learning difference i.e. prev_u - U goes below
           # this value break the learning
  
i=0
while(True):
  i += 1
  prev_u = np.copy(U)
  for s in range(env.observation_space.n):
    q_sa = [sum([p*(r + y*prev_u[s_]) for p, s_, r, _ in env.env.P[s][a]]) for a in range(env.action_space.n)]
    if s not in termS:
      U[s] = np.max(q_sa)
      
  if (np.sum(np.fabs(prev_u - U))<=eps):
    print('Value-iteration converged at iteration# {}'.format(i+1))
    break
    
print("After learning completion printing the utilities for each states below from state ids 0-15")
print()
print(U[:4])
print(U[4:8])
print(U[8:12])
print(U[12:16])

Value-iteration converged at iteration# 25
After learning completion printing the utilities for each states below from state ids 0-15

[0.023482   0.00999637 0.00437564 0.0023448 ]
[ 0.0415207  -1.         -0.19524141 -1.        ]
[ 0.09109598  0.20932556  0.26362693 -1.        ]
[-1.          0.43048408  0.97468581  1.        ]


What does this output mean? \\
The state representation is interpreted as following: \\
0 1 2 3 \\
4 5 6 7 \\
8 9 10 11 \\
12 13 14 15 \\
The starting state is $s=0$, then $U(s=0)=0.023482$, next there are 4 actions available: *UP, DOWN, LEFT, RIGHT*, \\
At, $s=0$, 
* action UP is taken, the $s\_new = 0$, therefore, $u[s\_new] =0.023482$
* action DOWN is taken, the $s\_new = 4$, therefore, $u[s\_new] = 0.0415207$
* action LEFT is taken, the $s\_new = 0$, therefore, $u[s\_new] =0.023482$
* action RIGHT is taken, the $s\_new = 1$, therefore, $u[s\_new] = 0.00999637$

Thus, max is $u(s\_new=4)-0.0415207$, therefore, the action taken is DOWN and $s\_new=4$.

Now at $=4$,
... [continue the same process]

Therefore, the final policy contains DOWN, DOWN, RIGHT, DOWN, RIGHT, and RIGHT to reach from $s=0$ (start state) to $s=15$ (goal state) by avoiding hole state (5, 7, 11, 12)